### ISSUE TITLE
Single custom Loss function for Multiple Output Regression not working

### ISSUE THREAD

**Issue by LouisAsh** [rstudio/keras#1216]
Hi and thanks for the amazing community around Keras!

**What I am trying to do:** create a single custom Loss function to be optimized by a Multiple Output Regression.

**Problem**: while my attempts at customizing loss functions for Multiple Output Regression do seem to be working, Keras still seems to be calling the customized function for each output. There is no specific documentation on how to do this properly in the package documentation - so I have no idea whether this is a bug.

**What have I tried**: I tried dozens of different way to specify the custom loss function. Also, I have extensively searched online for related issues. There is a sheer amount of questions on this topic online, but nearly all about debugging similar problems with Keras directly within Python ([example 1](https://github.com/keras-team/keras/issues/13340), [example 2](https://github.com/keras-team/keras/issues/13654)). The few questions I found online about that when using the keras package remain unanswered even after years ([example 1](https://stackoverflow.com/questions/49992346/neural-network-for-regression-using-keras-in-r), [example 2](https://stackoverflow.com/questions/49992346/neural-network-for-regression-using-keras-in-r)).

**Reproducible Example**:

First I create some fake data so we can have a minimally reproducible example.

```
library(keras)
n <- 10000
k <- 4
x <- runif(n, -100, 100)
data <- data.frame()
for(i in 1:k)
{
    data[1:n,i] <- x + runif(n, -0.01, 0.01)
}
y1 <- apply(data, 1, function(x) median(as.numeric(x)))
y2 <- apply(data, 1, function(x) sd(as.numeric(x)))
targets <- data.frame(y1=y1, y2=y2)

cases <- sample(1:nrow(data), round(0.5*nrow(data)))
train_targets <- targets[cases,]
test_targets <- targets[-cases,]
train_data <- data[cases, ]
test_data <- data[-cases, ]
mean <- apply(train_data, 2, mean)
std <- apply(train_data, 2, sd)
train_data <- scale(train_data, center = mean, scale = std)
test_data <- scale(test_data, center = mean, scale = std)
```

Then, let's try a rather silly custom loss function just for the sake of example:
```
custom_loss <- function(y_pred, y_true)
{
    result <- abs(y_pred[1] - y_true[1]) * abs(y_pred[2] - y_true[2])
    return(result)
}
```

Finally, let's define the model and run it:

```
build_model <- function()
{
    input_tensor <- layer_input(shape = dim(train_data)[[2]])
    base_model <- input_tensor %>%
                    layer_dense(units = 64, activation = "relu") %>%
                    layer_dense(units = 64, activation = "relu")
    y1_prediction <- base_model %>% layer_dense(units = 1, name="y1")
    y2_prediction <- base_model %>% layer_dense(units = 1, name="y2")
    model <- keras_model(input_tensor, list(y1_prediction, y2_prediction))
    # model %>% compile(optimizer = "rmsprop", loss = list(y1="mse", y2="mse"))
    model %>% compile(optimizer = "rmsprop", loss = custom_loss)
}

model <- build_model()
history <- model %>% fit(train_data, list(as.numeric(train_targets$y1),
                                          as.numeric(train_targets$y2)),
                         epochs = 2, batch_size = 256, verbose = 1)
```

Result: I still get two loss function results, one for each output:

> Epoch 1/200
20/20 [==============================] - 2s 3ms/step - loss: 2266.0321 - y1_loss: 2265.9965 - y2_loss: 0.0357
Epoch 2/200
20/20 [==============================] - 0s 5ms/step - loss: 1795.3417 - y1_loss: 1795.3154 - y2_loss: 0.0264

Comparing the above code example with solutions implemented directly in Python, I would think it should have worked as is. In which case, the model still optimizing two losses may be a bug. If there is a proper specification to work around that, please let me know how could I correctly specify the model such as to have it optimize one single custom loss function to which the multiple predicted and true values are inputs?

**Comment by t-kalinowski** [comment]
Hi, thanks for filing and welcome.

When you define a multi-output model, keras calculates and reports a loss for each output. This is intended. Though separate losses are reported for each of the outputs, notice there there is also a unspecified `loss` reported: this is the summation of the losses from each of the outputs. Keras calculates and reports the total loss and this is what is used by keras for the backprop + train_step.

I updated your example code into a minimal example of how you can supply custom losses for a multi output model and include it below.

A few notes:

* You can supply a named list of functions to `compile(loss = )`. The names should match what you provided to the argument `name = ` when constructing the final output tensors.
* If you supply only a single loss function, that same loss function gets recycled and called for each of the outputs. 
* I changed the shapes of the two output tensors to make this clear.

* In your `custom_loss` functions, the extract `[` call is maybe not working how you think it's working. When in doubt, you can see always see exactly what is supplied to any R function by calling `print` on the arguments, or even dropping into a debugger by including a `browser()` call in the function. If you call `print(y_true)`, you'll see that `y_true` is a tensor with `shape=(None, 1)`. The first dimension is the batch dimension (which is typically `None`, meaning it can be be any/unknown size). Indexing into the tensor with `y_pred[1]` is taking just the first row, aka, the first observation in the batch. Perhaps you mean `y_pred[,1]`? On the other hand, if you intended to only look at the first row for demonstration purposes, I would still recommend writing `y_pred[1,]` to be explicit about the shape.

* I suggest not coercing inputs supplied to `fit` with `as.numeric`. The `dim` attribute is the most reliable way to communicate your intended shape of a dataset to the framework, and if there is a mismatch between what you intend and what keras receives, you want it to fail early rather than introduce the risk of silent (unintended) recycling.

* Similarly, `data.frame` drops the `dim` attribute of inputs. If you want to use data frames to manage your arrays, I suggest using `tibble::tibble` here instead, because it preserves the `dim` attribute of array-columns and treats them correctly when subsetting the data frame. Though not essential to your question, I've also included how that can work in the example below.

```r
library(keras)
options(warn=1)

random_array <- function(...)
  array(runif(prod(c(...))), dim = c(...))

assert_shape <- function(tensor, shp)
  stopifnot(identical(dim(tensor), shp))

n <- 10000
k <- 4

x <- random_array(n, k)
y1 <- random_array(n, 1)
y2 <- random_array(n, 2)

data <- tibble::tibble(x, y1, y2)

cases <- sample.int(nrow(data), round(0.5 * nrow(data)))
train_data <- data[cases,]
test_data <- data[-cases,]

input <- layer_input(shape(4))

common_base <- input %>%
  layer_dense(units = 64, activation = "relu") %>%
  layer_dense(units = 64, activation = "relu")

output_y1 <- common_base %>% layer_dense(units = 1, name = "y1")
output_y2 <- common_base %>% layer_dense(units = 2, name = "y2")

model <- keras_model(input, list(output_y1, output_y2))

model %>%
  compile(
    optimizer = "rmsprop",
    loss = list(
      y1 = function(y_pred, y_true) {
        print("tracing y1 loss")
        # print(y_pred); print(y_true) # see what what tensor keras calls it with
        # browser() # you can also drop into a debugger here
        assert_shape(y_pred, shape(NULL, 1))
        k_sum(y_pred + y_true)
      },
      y2 = function(y_pred, y_true) {
        print("tracing y2 loss")
        # print(y_pred); print(y_true); browser()
        assert_shape(y_pred, shape(NULL, 2))
        k_sum(y_pred + y_true)
      }
    ))

history <- model %>%
  fit(train_data$x,
      list(y1 = train_data$y1,
           y2 = train_data$y2))
```

**Comment by LouisAsh** [comment]
@t-kalinowski: thanks for this example, very helpful to learn some specific details about how this works!

Now, if I understood correctly, you are saying that currently it is only possible to:

(1) pass a single custom loss function that will, then, be applied by Keras individually to each and all of the multiple outputs;
(2) pass a different custom loss function for each of the multiple outputs, (like in your code example above).

In both cases, then the sum of those individual output losses will be quantity that Keras will optimize for.

For example, let's supposed we have 4 outputs, y1, y2, y3 and y4. In case (1) above, it is possible to specify a single f(.) function that Keras will apply to each of those outputs, and then the quantity Keras will optimize is f(y1)+f(y2)+f(y3)+f(y4). On the other hand, what your example shows is that it is also possible to specify a separate different function for each output, in which case Keras will optimize f1(y1) + f2(y2) + f3(y3) + f4(y4).


However, what I would like to be able to do is different. Rather, I would like to:

(3) define a single custom loss function that instead of being applied to each individual output separately, would instead aggregate each output in a particular way. That aggregation would then be the quantity Keras optimizes.

Again in the 4 outputs example above, I would like to be able to specify a function f(y1, y2, y3, y4) for Keras to optimize.


Would it be possible? If not, then I certainly think it should be possible in the future - hence I can make this reply into a new issue as formal feature suggestion.

**Comment by t-kalinowski** [comment]
I think you can do this by defining a custom training loop. Please take a look at the guide here: https://keras.rstudio.com/articles/guides/customizing_what_happens_in_fit.html

As an alternative to defining a `train_step()` method, if `y1`, `y2`,`y3`,`y4` have compatible shapes, you could combine them with `layer_concatenate()` and then a single tensor gets passed into your loss function: `f(combined_tensors)`. `layer_reshape()` and/or `layer_flatten()` might be helpful for get shapes to match up.
